File: src/UserAccountModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import { UserAccountModuleEvents } from './UserAccount.types';

declare class UserAccountModule extends NativeModule<UserAccountModuleEvents> {
    /**
     * Query areas where the SMS verification code service is available.
     * Returns a comma-separated string of region codes.
     */
    getWhiteListWhoCanSendMobileCode(): Promise<string>;

    /**
     * Send a verification code to a mobile phone number.
     * @param userName - The mobile phone number.
     * @param region - The region code (from getWhiteListWhoCanSendMobileCode or regionList).
     * @param countryCode - The country code, e.g. "86".
     * @param type - 1: register, 2: login, 3: reset password.
     */
    sendVerifyCode(
        userName: string,
        region: string,
        countryCode: string,
        type: number
    ): Promise<void>;

    /**
     * Verify a verification code.
     * @param userName - The mobile phone number or email.
     * @param region - The region code. Pass empty string for default.
     * @param countryCode - The country code, e.g. "86".
     * @param code - The verification code received.
     * @param type - 1: register, 2: login, 3: reset password.
     */
    checkVerifyCode(
        userName: string,
        region: string,
        countryCode: string,
        code: string,
        type: number
    ): Promise<boolean>;

    /**
     * Register a new account with a mobile phone number.
     * Requires a verification code sent via sendVerifyCode with type 1.
     * @param countryCode - The country code, e.g. "86".
     * @param phoneNumber - The mobile phone number.
     * @param password - The password.
     * @param code - The verification code.
     */
    registerWithPhone(
        countryCode: string,
        phoneNumber: string,
        password: string,
        code: string
    ): Promise<void>;

    /**
     * Log in with a mobile phone number and password.
     * @param countryCode - The country code, e.g. "86".
     * @param phoneNumber - The mobile phone number.
     * @param password - The password.
     */
    loginWithPhone(
        countryCode: string,
        phoneNumber: string,
        password: string
    ): Promise<void>;

    /**
     * Log in with a mobile phone number and verification code.
     * Requires a verification code sent via sendVerifyCode with type 2.
     * @param phoneNumber - The mobile phone number.
     * @param countryCode - The country code, e.g. "86".
     * @param code - The verification code.
     */
    loginWithPhoneCode(
        phoneNumber: string,
        countryCode: string,
        code: string
    ): Promise<void>;

    /**
     * Reset the password for an account registered with a mobile phone number.
     * Requires a verification code sent via sendVerifyCode with type 3.
     * @param countryCode - The country code, e.g. "86".
     * @param phoneNumber - The mobile phone number.
     * @param newPassword - The new password.
     * @param code - The verification code.
     */
    resetPasswordWithPhone(
        countryCode: string,
        phoneNumber: string,
        newPassword: string,
        code: string
    ): Promise<void>;

    /**
     * Send a verification code to bind an email address to the current account.
     * @param email - The email address.
     * @param countryCode - The country code, e.g. "86".
     */
    sendBindingVerificationCodeWithEmail(
        email: string,
        countryCode: string
    ): Promise<void>;

    /**
     * Bind an email address to the current account.
     * @param email - The email address.
     * @param countryCode - The country code, e.g. "86".
     * @param code - The verification code.
     * @param sId - The session ID (from the logged-in user object).
     */
    bindEmail(
        email: string,
        countryCode: string,
        code: string,
        sId: string
    ): Promise<void>;

    // ─── Email-based registration and login ─────────────────────────────

    /**
     * Register a new account with an email address.
     * Requires a verification code sent via sendVerifyCode with type 1.
     * @param countryCode - The country code, e.g. "86".
     * @param email - The email address.
     * @param password - The password.
     * @param code - The verification code.
     */
    registerWithEmail(
        countryCode: string,
        email: string,
        password: string,
        code: string
    ): Promise<void>;

    /**
     * Log in with an email address and password.
     * @param countryCode - The country code, e.g. "86".
     * @param email - The email address.
     * @param password - The password.
     */
    loginWithEmail(
        countryCode: string,
        email: string,
        password: string
    ): Promise<void>;

    /**
     * Log in with an email address and verification code.
     * Requires a verification code sent via sendVerifyCode with type 2.
     * @param email - The email address.
     * @param countryCode - The country code, e.g. "86".
     * @param code - The verification code.
     */
    loginWithEmailCode(
        email: string,
        countryCode: string,
        code: string
    ): Promise<void>;

    /**
     * Reset the password for an account registered with an email address.
     * Requires a verification code sent via sendVerifyCode with type 3.
     * @param countryCode - The country code, e.g. "86".
     * @param email - The email address.
     * @param newPassword - The new password.
     * @param code - The verification code.
     */
    resetPasswordWithEmail(
        countryCode: string,
        email: string,
        newPassword: string,
        code: string
    ): Promise<void>;

    // ─── Phone binding ──────────────────────────────────────────────────

    /**
     * Send a verification code to bind a phone number to the current account.
     * @param countryCode - The country code, e.g. "86".
     * @param phoneNumber - The mobile phone number.
     */
    sendBindVerifyCodeWithPhone(
        countryCode: string,
        phoneNumber: string
    ): Promise<void>;

    /**
     * Bind a phone number to the current account.
     * @param countryCode - The country code, e.g. "86".
     * @param phoneNumber - The mobile phone number.
     * @param code - The verification code.
     */
    bindPhone(
        countryCode: string,
        phoneNumber: string,
        code: string
    ): Promise<void>;

    // ─── UID-based login ────────────────────────────────────────────────

    /**
     * Login or register with a UID from your own user system.
     * If the account exists, logs in. If not, registers and logs in.
     * @param countryCode - The country code, e.g. "86".
     * @param uid - The unique user identifier from your system.
     * @param password - A random identifier matching the account (not the user's actual password).
     * @param createHome - Whether to create a default home.
     */
    loginOrRegisterWithUid(
        countryCode: string,
        uid: string,
        password: string,
        createHome: boolean
    ): Promise<any>;

    // ─── Third-party login ──────────────────────────────────────────────

    /**
     * Login with a Facebook account.
     * @param countryCode - The country code, e.g. "86".
     * @param token - The token granted by Facebook.
     */
    loginWithFacebook(countryCode: string, token: string): Promise<void>;

    /**
     * Login with a Twitter account.
     * @param countryCode - The country code, e.g. "86".
     * @param key - The key granted by Twitter.
     * @param secret - The secret granted by Twitter.
     */
    loginWithTwitter(
        countryCode: string,
        key: string,
        secret: string
    ): Promise<void>;

    /**
     * Login with OAuth 2.0 (generic method for Apple, Google, LINE, etc.).
     * @param type - The provider type: "ap" (Apple), "gg" (Google), "line" (LINE), "tw" (Twitter), "wx" (WeChat), "qq" (QQ), "" (Tuya UID).
     * @param countryCode - The country code, e.g. "86".
     * @param accessToken - The token/idToken granted by the provider.
     * @param extraInfo - Additional info required by the provider (e.g. { pubVersion: 1 } for Google).
     */
    loginWithAuth2(
        type: string,
        countryCode: string,
        accessToken: string,
        extraInfo: Record<string, any>
    ): Promise<void>;

    /**
     * Login with a WeChat account.
     * @param countryCode - The country code, e.g. "86".
     * @param code - The code granted by WeChat.
     */
    loginWithWechat(countryCode: string, code: string): Promise<void>;

    /**
     * Login with a Tencent QQ account.
     * @param countryCode - The country code, e.g. "86".
     * @param userId - The QQ user ID (open ID).
     * @param accessToken - The access token granted by QQ.
     */
    loginWithQQ(
        countryCode: string,
        userId: string,
        accessToken: string
    ): Promise<void>;

    // ─── Anonymous account ──────────────────────────────────────────────

    /**
     * Register and login with an anonymous account.
     * @param countryCode - The country code, e.g. "1".
     * @param userName - A nickname for the anonymous account (e.g. device name).
     */
    registerAnonymous(countryCode: string, userName: string): Promise<void>;

    /**
     * Delete an anonymous account. The account is removed immediately.
     */
    deleteAnonymousAccount(): Promise<void>;

    /**
     * Bind a phone number or email to an anonymous account, converting it to a normal account.
     * Requires a verification code sent via sendVerifyCode first.
     * @param countryCode - The country code, e.g. "1".
     * @param userName - The phone number or email to bind.
     * @param code - The verification code.
     * @param password - The password to set for the account.
     */
    usernameBinding(
        countryCode: string,
        userName: string,
        code: string,
        password: string
    ): Promise<void>;

    // ─── Account management ─────────────────────────────────────────────

    /**
     * Set the temperature unit for the user.
     * @param tempUnit - 1 for °C, 2 for °F.
     */
    updateTempUnit(tempUnit: number): Promise<void>;

    /**
     * Update the user's nickname.
     * @param nickname - The new nickname.
     */
    updateNickname(nickname: string): Promise<void>;

    /**
     * Update the user's time zone.
     * @param timeZoneId - The time zone ID, e.g. "Asia/Shanghai".
     */
    updateTimeZone(timeZoneId: string): Promise<void>;

    /**
     * Update the user's location coordinates.
     * @param latitude - The latitude.
     * @param longitude - The longitude.
     */
    updateLocation(latitude: number, longitude: number): void;

    /**
     * Change the phone number or email bound to the account.
     * Requires a verification code sent beforehand.
     * @param account - The new phone number or email.
     * @param countryCode - The country code, e.g. "86".
     * @param code - The verification code.
     */
    changeBindAccount(
        account: string,
        countryCode: string,
        code: string
    ): Promise<void>;

    /**
     * Log out of the current account.
     */
    logout(): Promise<void>;

    /**
     * Deactivate/delete the current account.
     * The account is permanently deleted 7 days after this request
     * unless the user logs in again during that period.
     */
    cancelAccount(): Promise<void>;
}

export default requireNativeModule<UserAccountModule>('ExpoTuyaUserAccount');

File: ios/UserAccountModule.swift
import ExpoModulesCore
import ThingSmartBaseKit

public class UserAccountModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoTuyaUserAccount")

    Events("onSessionInvalid")

    // MARK: - Query areas for verification code service

    AsyncFunction("getWhiteListWhoCanSendMobileCode") { (promise: Promise) in
      ThingSmartUser.sharedInstance().getWhiteListWhoCanSendMobileCodeSuccess({ regions in
        promise.resolve(regions ?? "")
      }, failure: { error in
        promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to get whitelist regions"]))
      })
    }

    // MARK: - Send verification code

    AsyncFunction("sendVerifyCode") { (userName: String, region: String, countryCode: String, type: Int, promise: Promise) in
      ThingSmartUser.sharedInstance().sendVerifyCode(
        withUserName: userName,
        region: region,
        countryCode: countryCode,
        type: type,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to send verification code"]))
        }
      )
    }

    // MARK: - Verify verification code

    AsyncFunction("checkVerifyCode") { (userName: String, region: String, countryCode: String, code: String, type: Int, promise: Promise) in
      ThingSmartUser.sharedInstance().checkCode(
        withUserName: userName,
        region: region.isEmpty ? nil : region,
        countryCode: countryCode,
        code: code,
        type: type,
        success: { result in
          promise.resolve(result)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to verify code"]))
        }
      )
    }

    // MARK: - Register with phone

    AsyncFunction("registerWithPhone") { (countryCode: String, phoneNumber: String, password: String, code: String, promise: Promise) in
      ThingSmartUser.sharedInstance().register(
        byPhone: countryCode,
        phoneNumber: phoneNumber,
        password: password,
        code: code,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to register"]))
        }
      )
    }

    // MARK: - Login with phone and password

    AsyncFunction("loginWithPhone") { (countryCode: String, phoneNumber: String, password: String, promise: Promise) in
      ThingSmartUser.sharedInstance().login(
        byPhone: countryCode,
        phoneNumber: phoneNumber,
        password: password,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login"]))
        }
      )
    }

    // MARK: - Login with phone and verification code

    AsyncFunction("loginWithPhoneCode") { (phoneNumber: String, countryCode: String, code: String, promise: Promise) in
      ThingSmartUser.sharedInstance().login(
        withMobile: phoneNumber,
        countryCode: countryCode,
        code: code,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login with code"]))
        }
      )
    }

    // MARK: - Reset password

    AsyncFunction("resetPasswordWithPhone") { (countryCode: String, phoneNumber: String, newPassword: String, code: String, promise: Promise) in
      ThingSmartUser.sharedInstance().resetPassword(
        byPhone: countryCode,
        phoneNumber: phoneNumber,
        newPassword: newPassword,
        code: code,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to reset password"]))
        }
      )
    }

    // MARK: - Send binding verification code for email

    AsyncFunction("sendBindingVerificationCodeWithEmail") { (email: String, countryCode: String, promise: Promise) in
      ThingSmartUser.sharedInstance().sendBindingVerificationCode(
        withEmail: email,
        countryCode: countryCode,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to send binding verification code"]))
        }
      )
    }

    // MARK: - Bind email

    AsyncFunction("bindEmail") { (email: String, countryCode: String, code: String, sId: String, promise: Promise) in
      ThingSmartUser.sharedInstance().bindEmail(
        email,
        withCountryCode: countryCode,
        code: code,
        sId: sId,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to bind email"]))
        }
      )
    }

    // MARK: - Register with email

    AsyncFunction("registerWithEmail") { (countryCode: String, email: String, password: String, code: String, promise: Promise) in
      ThingSmartUser.sharedInstance().register(
        byEmail: countryCode,
        email: email,
        password: password,
        code: code,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to register with email"]))
        }
      )
    }

    // MARK: - Login with email and password

    AsyncFunction("loginWithEmail") { (countryCode: String, email: String, password: String, promise: Promise) in
      ThingSmartUser.sharedInstance().login(
        byEmail: countryCode,
        email: email,
        password: password,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login with email"]))
        }
      )
    }

    // MARK: - Login with email and verification code

    AsyncFunction("loginWithEmailCode") { (email: String, countryCode: String, code: String, promise: Promise) in
      ThingSmartUser.sharedInstance().login(
        withEmail: email,
        countryCode: countryCode,
        code: code,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login with email code"]))
        }
      )
    }

    // MARK: - Reset password with email

    AsyncFunction("resetPasswordWithEmail") { (countryCode: String, email: String, newPassword: String, code: String, promise: Promise) in
      ThingSmartUser.sharedInstance().resetPassword(
        byEmail: countryCode,
        email: email,
        newPassword: newPassword,
        code: code,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to reset password with email"]))
        }
      )
    }

    // MARK: - Send binding verification code for phone

    AsyncFunction("sendBindVerifyCodeWithPhone") { (countryCode: String, phoneNumber: String, promise: Promise) in
      ThingSmartUser.sharedInstance().sendBindVerifyCode(
        countryCode,
        phoneNumber: phoneNumber,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to send phone binding verification code"]))
        }
      )
    }

    // MARK: - Bind phone number

    AsyncFunction("bindPhone") { (countryCode: String, phoneNumber: String, code: String, promise: Promise) in
      ThingSmartUser.sharedInstance().mobileBinding(
        countryCode,
        phoneNumber: phoneNumber,
        code: code,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to bind phone number"]))
        }
      )
    }

    // MARK: - Login or register with UID

    AsyncFunction("loginOrRegisterWithUid") { (countryCode: String, uid: String, password: String, createHome: Bool, promise: Promise) in
      ThingSmartUser.sharedInstance().loginOrRegister(
        withCountryCode: countryCode,
        uid: uid,
        password: password,
        createHome: createHome,
        success: { result in
          promise.resolve(result)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login or register with UID"]))
        }
      )
    }

    // MARK: - Login with Facebook

    AsyncFunction("loginWithFacebook") { (countryCode: String, token: String, promise: Promise) in
      ThingSmartUser.sharedInstance().login(
        byFacebook: countryCode,
        token: token,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login with Facebook"]))
        }
      )
    }

    // MARK: - Login with Twitter

    AsyncFunction("loginWithTwitter") { (countryCode: String, key: String, secret: String, promise: Promise) in
      ThingSmartUser.sharedInstance().login(
        byTwitter: countryCode,
        key: key,
        secret: secret,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login with Twitter"]))
        }
      )
    }

    // MARK: - Login with OAuth 2.0

    AsyncFunction("loginWithAuth2") { (type: String, countryCode: String, accessToken: String, extraInfo: [String: Any], promise: Promise) in
      ThingSmartUser.sharedInstance().loginByAuth2(
        withType: type,
        countryCode: countryCode,
        accessToken: accessToken,
        extraInfo: extraInfo,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login with OAuth 2.0"]))
        }
      )
    }

    // MARK: - Login with WeChat

    AsyncFunction("loginWithWechat") { (countryCode: String, code: String, promise: Promise) in
      ThingSmartUser.sharedInstance().login(
        byWechat: countryCode,
        code: code,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login with WeChat"]))
        }
      )
    }

    // MARK: - Login with QQ

    AsyncFunction("loginWithQQ") { (countryCode: String, userId: String, accessToken: String, promise: Promise) in
      ThingSmartUser.sharedInstance().login(
        byQQ: countryCode,
        userId: userId,
        accessToken: accessToken,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to login with QQ"]))
        }
      )
    }

    // MARK: - Register anonymous account

    AsyncFunction("registerAnonymous") { (countryCode: String, userName: String, promise: Promise) in
      ThingSmartUser.sharedInstance().registerAnonymous(
        withCountryCode: countryCode,
        userName: userName,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to register anonymous account"]))
        }
      )
    }

    // MARK: - Delete anonymous account

    AsyncFunction("deleteAnonymousAccount") { (promise: Promise) in
      ThingSmartUser.sharedInstance().deleteAnonymousAccount(success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to delete anonymous account"]))
      })
    }

    // MARK: - Bind username to anonymous account

    AsyncFunction("usernameBinding") { (countryCode: String, userName: String, code: String, password: String, promise: Promise) in
      ThingSmartUser.sharedInstance().usernameBinding(
        withCountryCode: countryCode,
        userName: userName,
        code: code,
        password: password,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to bind username"]))
        }
      )
    }

    // MARK: - Set temperature unit

    AsyncFunction("updateTempUnit") { (tempUnit: Int, promise: Promise) in
      ThingSmartUser.sharedInstance().updateTempUnit(withTempUnit: tempUnit, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to update temperature unit"]))
      })
    }

    // MARK: - Update nickname

    AsyncFunction("updateNickname") { (nickname: String, promise: Promise) in
      ThingSmartUser.sharedInstance().updateNickname(nickname, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to update nickname"]))
      })
    }

    // MARK: - Update time zone

    AsyncFunction("updateTimeZone") { (timeZoneId: String, promise: Promise) in
      ThingSmartUser.sharedInstance().updateTimeZone(withTimeZoneId: timeZoneId, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to update time zone"]))
      })
    }

    // MARK: - Update location

    Function("updateLocation") { (latitude: Double, longitude: Double) in
      ThingSmartSDK.sharedInstance().updateLatitude(latitude, longitude: longitude)
    }

    // MARK: - Change bound account

    AsyncFunction("changeBindAccount") { (account: String, countryCode: String, code: String, promise: Promise) in
      ThingSmartUser.sharedInstance().changBindAccount(
        account,
        countryCode: countryCode,
        code: code,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to change bound account"]))
        }
      )
    }

    // MARK: - Logout

    AsyncFunction("logout") { (promise: Promise) in
      ThingSmartUser.sharedInstance().loginOut({
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to logout"]))
      })
    }

    // MARK: - Cancel (delete) account

    AsyncFunction("cancelAccount") { (promise: Promise) in
      ThingSmartUser.sharedInstance().cancelAccount({
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? NSError(domain: "ExpoTuyaUserAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to cancel account"]))
      })
    }

    // MARK: - Session expiration listener

    OnStartObserving {
      self.sessionObserver = NotificationCenter.default.addObserver(
        forName: NSNotification.Name(rawValue: "ThingSmartUserNotificationUserSessionInvalid"),
        object: nil,
        queue: .main
      ) { [weak self] _ in
        self?.sendEvent("onSessionInvalid")
      }
    }

    OnStopObserving {
      if let observer = self.sessionObserver {
        NotificationCenter.default.removeObserver(observer)
        self.sessionObserver = nil
      }
    }
  }

  private var sessionObserver: NSObjectProtocol?
}
